File: szavazas-core/src/main/java/hu/kdea/szavazas/ballotprocessor/qr/preprocess/ContrastEnhancementStep.java
package hu.kdea.szavazas.ballotprocessor.qr.preprocess;

import boofcv.struct.image.GrayU8;

public class ContrastEnhancementStep implements PreprocessingStep {
    @Override
    public GrayU8 apply(GrayU8 input) {
        int min = 255;
        int max = 0;
        for (int y = 0; y < input.height; y++) {
            for (int x = 0; x < input.width; x++) {
                int value = input.get(x, y);
                if (value < min) {
                    min = value;
                }
                if (value > max) {
                    max = value;
                }
            }
        }
        GrayU8 out = new GrayU8(input.width, input.height);
        if (max <= min) {
            return input.clone();
        }
        double scale = 255.0 / (max - min);
        for (int y = 0; y < input.height; y++) {
            for (int x = 0; x < input.width; x++) {
                int value = input.get(x, y);
                int adjusted = (int) ((value - min) * scale);
                out.set(x, y, Math.max(0, Math.min(255, adjusted)));
            }
        }
        return out;
    }
}

File: test-utils/src/main/java/hu/kdea/szavazas/ballotprocessor/qr/preprocess/ContrastEnhancementStep.kt
package hu.kdea.szavazas.ballotprocessor.qr.preprocess

import boofcv.struct.image.GrayU8

class ContrastEnhancementStep : PreprocessingStep {
    override val name = "contrast"
    override fun process(input: GrayU8): GrayU8 {
        var min = 255
        var max = 0
        for (y in 0 until input.height) {
            for (x in 0 until input.width) {
                val v = input.get(x, y).toInt()
                if (v < min) min = v
                if (v > max) max = v
            }
        }
        val out = GrayU8(input.width, input.height)
        if (max <= min) return input.clone()  // avoid division by zero
        val scale = 255.0 / (max - min)
        for (y in 0 until input.height) {
            for (x in 0 until input.width) {
                val v = input.get(x, y).toInt()
                out.set(x, y, ((v - min) * scale).toInt().coerceIn(0, 255))
            }
        }
        return out
    }
}